src/main/java/jcifs/smb/SmbFile.java

```diff
@@ -657,30 +659,72 @@ SmbFileHandleImpl openUnshared ( String uncPath, int flags, int access, int shar
+                        req.setCreateDisposition(Smb2CreateRequest.FILE_OVERWRITE_IF);
+                    }
+                    else if ( ( flags & SmbConstants.O_TRUNC ) == O_TRUNC ) {
+                        req.setCreateDisposition(Smb2CreateRequest.FILE_OVERWRITE);
+                    }
+                    else if ( ( flags & SmbConstants.O_EXCL ) == O_EXCL ) {
+                        req.setCreateDisposition(Smb2CreateRequest.FILE_CREATE);
+                    }
+                    else if ( ( flags & SmbConstants.O_CREAT ) == O_CREAT ) {
+                        req.setCreateDisposition(Smb2CreateRequest.FILE_OPEN_IF);
+                    }
+                    else {
+                        req.setCreateDisposition(Smb2CreateRequest.FILE_OPEN);
+                    }
+
+                    req.setShareAccess(sharing);
+                    req.setFileAttributes(attrs);
+                    Smb2CreateResponse resp = h.send(req);
+                    info = resp;
+                    fileSize = resp.getEndOfFile();
+                    fh = new SmbFileHandleImpl(config, resp.getFileId(), h, uncPath, flags, access, 0, 0, resp.getEndOfFile());
                 }
-                else {
-                    req.setCreateDisposition(Smb2CreateRequest.FILE_OPEN);
+                catch ( CIFSException | RuntimeException e ) {
+                    Smb2CreateResponse resp = req.getResponse();
+
+                    // We hit a symbolic link, parse the error data and resend for the 'real' directory or file path
+                    if (resp != null && resp.isReceived() && resp.getStatus() == NtStatus.NT_STATUS_STOPPED_ON_SYMLINK) {
+                        this.isSymlink = true;
+
+                        try {
+                            byte[] errorData = resp.getErrorData();
+                            Smb2ErrorContextResponse ecr = new Smb2ErrorContextResponse();
+                            int symLinkLength = ecr.readErrorContextResponse(errorData);
+                            log.debug("symLinkLength -> {}", symLinkLength);
+
+                            String substituteName = ecr.getSubstituteName();
+                            log.debug("substituteName -> {}", substituteName);
+
+                            int i = substituteName.indexOf(this.getShare());
+                            String realPath = substituteName.substring(i + this.getShare().length());
+                            log.debug("symLinkPath -> {}", uncPath);
+                            log.debug("realPath -> {}", realPath);
+
+                            if (this.isFile() && !uncPath.endsWith("\\")) { // is this a file in a symlink directory?
+                                String fileName = uncPath.substring(uncPath.lastIndexOf("\\") + 1);
+                                log.debug("fileName -> {}", fileName);
+                                return openUnshared (realPath + fileName, flags, access, sharing, attrs, options);
```

Comment:
As far as I have tested it does work for a symlink directory, I have been unit testing that use case but I will review.

Reply:
Works as designed for a symlink directory

Reply:
I still believe we should construct the symlink target path according to the spec and not using a heuristic locating the share name inside the UNC paths (but afterwards check that the share is still the same).
As per "2.2.2.2.1.1 Handling the Symbolic Link Error Response" one should construct the unparsed path component (i.e. the path below the symlink) from the last _UnparsedPathLength_ bytes of the request UNC path and append that to the provided _SubstituteName_.

Reply:
I've been reading those specs over and over and I still can't see how your code is going to handle a target path for a Substitute Name such as `\??\D:\GWTEST\Finance\smbSource\subSmbSource\` in SmbFile for a symlink directory with the original request of `SmbFile:659 - openUnshared: \Finance\symlinkTestDir\ flags: 0 access: 1 attrs: 80 options: 0` without doing what I have already coded to extract the root path to resolve the target path.

In this case unparsedPathLength = 0, so what am I expected to extract for the unparsed path?

`2022-10-06 21:51:58 INFO  SymLinkTest:114 - Resource [symlinkTestDir/] is a directory.`
`2022-10-06 21:51:58 INFO  SymLinkTest:124 - Resource [symlinkTestDir/] is not a file.`
`2022-10-06 21:51:58 DEBUG SmbFile:852 - queryPath: \Finance\symlinkTestDir\`
`2022-10-06 21:51:58 INFO  SymLinkTest:128 - Resource [symlinkTestDir/] is a symbolic link.`
`2022-10-06 21:51:58 DEBUG SmbFile:659 - openUnshared: \Finance\symlinkTestDir\ flags: 0 access: 1 attrs: 80 options: 0`
`2022-10-06 21:51:58 DEBUG SmbFile:784 - SymLink Path -> \Finance\symlinkTestDir\`
`2022-10-06 21:51:58 INFO  Smb2ErrorDataFormat:78 - unparsedPathLength -> 0`
`2022-10-06 21:51:58 DEBUG SmbFile:789 - SymLink Length -> 196`
`2022-10-06 21:51:58 DEBUG SmbFile:791 - Absolute Path -> true`
`2022-10-06 21:51:58 DEBUG SmbFile:792 - Print Name -> D:\GWTEST\Finance\smbSource\subSmbSource\`
`2022-10-06 21:51:58 DEBUG SmbFile:795 - Substitute Name -> \??\D:\GWTEST\Finance\smbSource\subSmbSource\`
`2022-10-06 21:51:58 DEBUG SmbFile:801 - Root Path -> \Finance\smbSource\subSmbSource\`
`2022-10-06 21:51:58 DEBUG SmbFile:825 - Target Path -> \Finance\smbSource\subSmbSource\`
`2022-10-06 21:51:58 DEBUG SmbFile:659 - openUnshared: \Finance\smbSource\subSmbSource\ flags: 0 access: 1 attrs: 80 options: 0`

Reply:
Judging from the first clarification section in 2.2.2.2.1.1, I don't think that an absolute symlink to a local drive should be resolved at all in that way:

> An absolute symbolic link located on the server links "\\MachX\ShareY\Public\ProtocolDocs" to "\??\D:\DonHall\MiscDocuments\PDocs".
> The original open request is for "\\MachX\ShareY\Public\ProtocolDocs\DailyDocs\[MS-SMB].doc".
> The unparsed portion of the path name will be "\DailyDocs\[MS-SMB].doc". Appending the substitute name with the unparsed portion of the file name gives the new target path name of "\??\D:\DonHall\MiscDocuments\PDocs\DailyDocs\[MS-SMB].doc".

Therefore that symlink would point to the local D: drive _on the client_ (which I don't think we would want to handle at all), not transparently to another server-side resource. A different story might be a absolute UNC path, which might point to the same share (this could be easily handled), or to another share or server (this would be much more difficult, as mentioned before)